File: backend/controllers/auth_controller.py
from fastapi import HTTPException, Header
from typing import Optional

from dtos.auth_dto import LoginDTO, TokenResponseDTO, TokenValidationDTO
from services.auth_service import AuthService


class AuthController:
    """Controller layer for authentication request handling"""
    
    def __init__(self, auth_service: AuthService):
        self.auth_service = auth_service
    
    async def login(self, login_dto: LoginDTO) -> TokenResponseDTO:
        """Handle login request"""
        try:
            token_response = await self.auth_service.login(login_dto)
            return token_response
        except ValueError as e:
            raise HTTPException(status_code=401, detail=str(e))
        except Exception as e:
            raise HTTPException(status_code=500, detail="Internal server error")
    
    async def validate_token(self, authorization: Optional[str] = Header(None)) -> TokenValidationDTO:
        """Handle token validation request"""
        if not authorization or not authorization.startswith("Bearer "):
            return TokenValidationDTO(valid=False, user=None)
        
        token = authorization.replace("Bearer ", "")
        
        try:
            user = await self.auth_service.validate_token(token)
            if user:
                return TokenValidationDTO(valid=True, user=user)
            return TokenValidationDTO(valid=False, user=None)
        except Exception:
            return TokenValidationDTO(valid=False, user=None)
    
    async def logout(self) -> dict:
        """Handle logout request"""
        return {"message": "Logged out successfully"}

File: backend/src/controllers/auth.controller.ts
import { Request, Response } from 'express';
import { AuthService } from '../services/auth.service';
import { LoginDTO, TokenValidationDTO } from '../dtos/auth.dto';

export class AuthController {
  private authService: AuthService;

  constructor() {
    this.authService = new AuthService();
  }

  async login(req: Request, res: Response): Promise<void> {
    try {
      const loginDto = req.body as LoginDTO;
      const tokenResponse = await this.authService.login(loginDto);
      res.status(200).json(tokenResponse);
    } catch (error) {
      if (error instanceof Error && error.message === 'Invalid email or password') {
        res.status(401).json({ detail: error.message });
      } else {
        res.status(500).json({ detail: 'Internal server error' });
      }
    }
  }

  async validateToken(req: Request, res: Response): Promise<void> {
    try {
      const authHeader = req.headers.authorization;

      if (!authHeader || !authHeader.startsWith('Bearer ')) {
        const response: TokenValidationDTO = { valid: false, user: undefined };
        res.status(200).json(response);
        return;
      }

      const token = authHeader.replace('Bearer ', '');
      const user = await this.authService.validateToken(token);

      if (user) {
        const response: TokenValidationDTO = { valid: true, user };
        res.status(200).json(response);
      } else {
        const response: TokenValidationDTO = { valid: false, user: undefined };
        res.status(200).json(response);
      }
    } catch (error) {
      const response: TokenValidationDTO = { valid: false, user: undefined };
      res.status(200).json(response);
    }
  }

  async logout(req: Request, res: Response): Promise<void> {
    res.status(200).json({ message: 'Logged out successfully' });
  }
}
